### ISSUE TITLE
yarn/node execution permission denied

### ISSUE THREAD

**Issue by mhermosi** [eirslett/frontend-maven-plugin#743]
Hi, 

We are using frontend-maven-plugin v1.6 on a Bitbucket pipeline using a custom docker image.
this issue has appeared only on that environment if I run the maven build locally it does not fail.

These are part of the maven output:

**Comment by mhermosi** [comment]
Is this project abandoned?

**Comment by tubbynl** [comment]
dunnow, i'm having the same issue running from within a `Jenkinsfile`

```
[INFO] --- frontend-maven-plugin:1.6:yarn (default-cli) @ nenhkz ---
[INFO] Running 'yarn ' in /var/lib/jenkins/workspace/kz_feature_SPX-450-jhipster-JADFRK7UKVPT6UEHS4NQ2N7P32P64OS7524VZILJY7QQYG4RMUEQ
[ERROR] /var/lib/jenkins/workspace/..../node/yarn/dist/bin/yarn: exec: line 32: node: Permission denied
```

i've already tried `sh "chmod +x node/yarn/dist/bin/yarn"`

**Comment by mhermosi** [comment]
Hi @tubbynl  I think problems may be different, at least the cookie thing. Here seems to be an execution permission problem, so maybe developer of this plugin should verify and update execution permission of installed tools like Yarn or Bower etc... I have forked this thing and I will try to add a method to force `mode 0777` on `INSTALL_DIR/dist/bin/yarn` and see how it goes.

**Comment by mhermosi** [comment]
The other problem may be a nodejs permission problem:

**Comment by tubbynl** [comment]
i ended up creating a build container  having maving+yarn (alpine now has a yarn package) and not using the `install-node-and-yarn`

**Comment by mhermosi** [comment]
@tubbynl unfortunately for us. for local environment build we need consistency with the tools. so we have to keep using this.

**Comment by tubbynl** [comment]
for what it's worth then; 
* locally on macOS it works
* on Jenkins running in a Jenkinsfile using the `maven:3.5.4-alpine` image it breaks with the above permission error

The build image i created is 
```
FROM maven:3.5.4-alpine
RUN apk add yarn
```

**Comment by tubbynl** [comment]
ok, after fiddling i discovered that the `Jenkinsfile` build using `install-node-and-yarn` did succeed if i use `maven:3.5.4` (non-alpine) as build image

**Comment by majestica** [comment]
@tubbynl can you share your solution?

**Comment by narimanmandi** [comment]
i have the same issue 
everything is all right when i run my application in local(mac os)
but build stage fails with the same error while trying to build via gitlab ci/cd pipelines
@tubbynl do you have any suggestion ?

**Comment by tubbynl** [comment]
my current "solution" is this bitbucket-pipelines.yml file

```yarn
pipelines:
    default:
    - step:
          name: yarn test
          image: node:10-alpine
          caches:
          - node
          script:
          - printenv
          - yarn install --frozen-lockfile
          - yarn test
    - step:
          name: mvn test
          image: maven:3.5-jdk-8-alpine
          caches:
          - maven
          script:
          - mvn -v
          - mvn -B clean test
    tags:
        '*':
        - step:
              name: yarn build
              image: node:10-alpine
              caches:
              - node
              script:
              - yarn install --frozen-lockfile
              - yarn test
              - yarn build
              artifacts:
              - target/**
        - step:
              name: mvn build
              image: maven:3.5-jdk-8-alpine
              caches:
              - maven
              - node
              script:
              - mvn -B versions:set -DnewVersion=$BITBUCKET_TAG
              - mvn -B package
              artifacts:
              - target/*.war
        - step:
              name: docker
              services:
              - docker
              script:
              - cp -R src/main/docker target/
              - cp target/*$BITBUCKET_TAG.war target/docker/app.war
              - docker version
              - docker login -u $DOCKER_USERNAME -p $DOCKER_PASSWORD $DOCKER_REPO
              - docker build target/docker/ -t $DOCKER_IMAGE:$BITBUCKET_TAG
              - docker push $DOCKER_IMAGE:$BITBUCKET_TAG
```

essentially; not using this front-end maven plugin when building on the CI; using standard node/maven alpines images to build the frontend and the backend and making sure the front-end build result is correctly included in the backend packaging structure.

**Comment by mhermosi** [comment]
yeah, i've decided to fork this project and try to see if I can get a solution.. it is simply not making any sense having a permission issue running on bitbucket.. I think basically is missing the execution flag I think let's see where this drives me.

**Comment by DonAscon** [comment]
@mhermosi  did you have any luck?
Having the same problem. I am unable to bring my vaadin software in production mode.

**Comment by mhermosi** [comment]
Hi, I was able to make this work, I am not sure if this may apply to everybody. but debugging my docker image based on alpine I realized that the error is not related with execution permission nor with yarn. it is node distribution itself. 

the plugin downloads node distribution which is fine. but in case of alpine linux, not sure if all other Linux images for docker (have not tried) as they are minimal linux distribution, some dependencies (shared libraries) are not available, that is the problem. So how did I overcome this was to create an image including the nodejs / npm distribution from alpine (in my case with node 8.12.0) and match that version with the one used with the plugin. once I did that, the shared libraries were present and the build worked fine.

maybe we should ask the people on node to create a static linked version of their distribution to be used with this library, because finally the plugin itself is not the culprit on this. unfortunately I was not able to find which libraries are missing because alpine does not have ldd and I have not had the time to dig on it to publish a list of missing shared libraries.

The example I saw with pipelines installing yarn and having node-10-alpine I think should work....

**Comment by mhermosi** [comment]
Oh also be aware that using nodejs on bitbucket (not sure if other CI tools) they set an env var called CI to true. that will mess your node build because CI = true will trigger automatically warning to stop the build as error do.

**Comment by DonAscon** [comment]
Thank you for youre quick reply. The pipeline mentioned above did not get me much further.
By installing node and yarn before i execute the maven plugin i managed to get yarn to work.

But the maven plugin is still looking for the locally installed node and can not find it. Which is kind of strange since i added a symlink to the previous installed node.

This is driving me nuts :/ Unable to bring my application in production mode.

**Comment by mhermosi** [comment]
it will keep looking for the local installed node, not the system installed one. the sole purpose of installing (on alpine using apk add nodejs and apk install nodejs-npm) is to get the library dependencies

**Comment by DonAscon** [comment]
Still having problems after installing both it seems.

I receive the following error: Transpilation with gulp has failed: 'gulp build_es6' failed. java.io.IOException: Cannot run program "/opt/atlassian/pipelines/agent/build/target/node/node" (in directory "/opt/atlassian/pipelines/agent/build/target"): error=2, No such file or directory

(Yarn did succeed)

I did install node in my pipeline:
  default:
  - step:
      image: maven:3.5-jdk-8-alpine
      name: Build
      caches:
      - maven
      script:
      - apk update
      - apk add --update nodejs nodejs-npm && node -v
      - which node
      - npm install -g yarn
      - yarn --version
      - which yarn
      #- docker info
      - mvn -e -Pproduction verify package  # -B batch mode makes Maven less verbose
      services:
        - docker

And the maven plugin from the last step also states:
[INFO] Installed node locally.

**Comment by mhermosi** [comment]
mmm this means your install is working both node and yarn (yarn used node to run) so rest is another problem with your build

**Comment by DonAscon** [comment]
Yes gulp doesn't matter here. What i do not understand, that the node binary is not found - although node was installed by the plugin.

**Comment by dveijck** [comment]
@mhermosi en @DonAscon not sure if it is still relevant for you but just posting it anyway for other people who might end up here.

I ran into the same issue and found that there is a similar issue #633 which contains the answer to the problem.

Basically the thing has to do with glibc which is quite a bit of a hassle on Alpine.
What fixed it for me was adding:

```
    apk --no-cache add ca-certificates wget && \
    wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub && \
    wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.28-r0/glibc-2.28-r0.apk && \
    apk add glibc-2.28-r0.apk && \
    wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.28-r0/glibc-bin-2.28-r0.apk && \
    apk add glibc-bin-2.28-r0.apk && \
    wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.28-r0/glibc-i18n-2.28-r0.apk && \
    apk add glibc-i18n-2.28-r0.apk && \
```

You will likely encounter messages like:

```
[ERROR] node: /usr/lib/libstdc++.so.6: no version information available (required by node)
```

but it seems these can be ignored.

It is definitely worth looking at [the alpine-pkg-glibc repository](https://github.com/sgerrand/alpine-pkg-glibc) as it contains some more detailed information and the latest available releases.

Hope it helps.

**Comment by breun** [comment]
The issue is that a yarn binary that is built to be used with glibc doesn't work on Alpine out of the box, because Alpine uses musl instead of glibc as its C standard library.

While you can add glibc to Alpine, I think it kind of defeats the point of using Alpine Linux, because it adds so much disk space usage to that otherwise tiny distro. You might be better off using a glibc-based Linux distribution instead (Debian Slim or something).

Or, frontend-maven-plugin needs to be able to download a musl-based Yarn binary, like it already does for NodeJS (see https://github.com/eirslett/frontend-maven-plugin/pull/853).
